src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionExecutor.cs

```diff
@@ -28,6 +28,32 @@ public static class ControllerActionExecutor
             return actionMethodExecutor.ExecuteAsync(instance, orderedActionArguments);
         }
 
+        public static async Task<TResult> ExecuteAsync<TResult>(
+            ObjectMethodExecutor actionMethodExecutor,
+            object instance,
+            object[] orderedActionArguments,
+            Func<ObjectMethodExecutor, object, TResult> createActionResult
+            )
+        {
+            var returnType = actionMethodExecutor.MethodInfo.ReturnType;
+
+            if (actionMethodExecutor.CanExecuteInSync)
+            {
+                var result = actionMethodExecutor.Execute(instance, orderedActionArguments);
+                if (returnType == typeof(Task))
+                {
+                    await (Task)result;
+                }
+
+                return createActionResult(actionMethodExecutor, result);
```

Comment:
This is still going to result in allocating a `Task<T>` when the action method is synchronous. Maybe this would be a good place to use `ValueTask<T>`